snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,387 @@
+        # want the first message to be consumed after a rebalance to be *the
+        # same* as the most recently consumed message. To account for this
+        # discrepancy, we disable the default automatic offset storage
+        # behavior, and use our own implementation that stores the offset of
+        # the most recently received message in a partition, instead of that
+        # messages's offset + 1.
+        self.__consumer = KafkaConsumer(
+            {**configuration, "enable.auto.offset.store": "false"}
+        )
+
+        self.__auto_offset_reset = get_enum_configuration_value(
+            configuration,
+            "auto.offset.reset",
+            {
+                "earliest": set(["smallest", "beginning"]),
+                "latest": set(["largest", "end"]),
+                "error": set(),
+            },
+            "largest",
+        )
+
+        # XXX: The only way to store offsets (and correspondingly, cause those
+        # stored offsets to be committed) is to use the automatic offset
+        # storage. (There is no ``store_offsets`` method.) That makes this
+        # configuration value effectively required
+        self.__enable_auto_offset_store = get_bool_configuration_value(
+            configuration, "enable.auto.offset.store", True
+        )
+
+    def __get_committed_offsets(
+        self, partitions: Sequence[TopicPartition]
+    ) -> Iterator[TopicPartition]:
+        """
+        Fetch the committed offset for each partition provided. If there is
+        no committed offset for a partition, the offset corresponding to the
+        behavior defined by the ``auto.offset.reset`` configuration variable
+        is used.
+        """
+        for partition in self.__consumer.committed(partitions):
+            if partition.error is not None:
+                raise partition.error
+
+            if not partition.offset == OFFSET_INVALID:
+                yield partition
+            else:
+                if self.__auto_offset_reset == "error":
+                    raise Exception  # TODO: Add error message here.
+
+                earliest, latest = self.__consumer.get_watermark_offsets(
+                    partition, cached=False
+                )
+                if self.__auto_offset_reset == "earliest":
+                    yield TopicPartition(partition.topic, partition.partition, earliest)
+                elif self.__auto_offset_reset == "latest":
+                    yield TopicPartition(partition.topic, partition.partition, latest)
+                else:
+                    raise ValueError(
+                        f'unexpected value for "auto.offset.reset": {self.__auto_offset_reset!r}'
+                    )
+
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+        on_revoke: Optional[Callable[[Sequence[TopicPartition]], Any]] = None,
+    ) -> None:
+        def on_assign_callback(
```

Comment:
I'd consider moving the part of the comment that explains the rebalance problem here (as docstring of the subscribe method). It would be more contextual and reduce the size of that wall of text.

Reply:
Interesting point, I'll see if I can make that work.

Reply:
Moved and slightly reworded that comment with aaef1c6c3e648531132743f695a6d14e80baf9e4.

Reply:
Moved and slightly reworded that comment with aaef1c6c3e648531132743f695a6d14e80baf9e4.